text/0000-dyn-error-generic-member-access.md

```diff
@@ -0,0 +1,335 @@
+- Feature Name: Add fns for generic member access to dyn Error and the Error trait
+- Start Date: 2020-04-01
+- RFC PR: [rust-lang/rfcs#0000](https://github.com/rust-lang/rfcs/pull/2895)
+- Rust Issue: [rust-lang/rust#0000](https://github.com/rust-lang/rust/issues/0000)
+
+# Summary
+[summary]: #summary
+
+This RFC proposes a pair of additions to the `Error` trait to support accessing
+generic forms of context from `dyn Error` trait objects, one method on the
+`Error` trait itself for returning references to members based on a given type
+id, and another fn implemented for `dyn Error` that uses a generic return type
+to get the type id to pass into the trait object's fn. These functions will act
+as a generalized version of `backtrace` and `source`, and would primarily be
+used during error reporting when rendering a chain of opaque errors.
```

Comment:
```suggestion
This RFC proposes two additions to the `Error` trait to support accessing
generic forms of context from `dyn Error` trait objects, generalizing the
pattern used in `backtrace` and `source` and allowing ecosystem iteration
on error reporting infrastructure outside of the standard library. The two
proposed additions are a new trait method `Error::provide_context` which
offers `TypeId`-based member lookup and a new inherent fn
`<dyn Error>::context` which makes use of an implementor's `provide_context`
to return a typed reference directly.

These additions would primarily be useful in "error reporting" contexts where we
typically no longer have type information and may be composing errors from many
sources.
```

IMO the definitions themselves are very readable as an "intro" to what you're proposing if you put them together:

```rust
pub trait Error {
    /// Provide an untyped reference to a member whose type matches the provided `TypeId`.
    ///
    /// Returns `None` by default, implementors are encouraged to override.
    fn provide_context(&self, ty: TypeId) -> Option<&dyn Any> {
        None
    }
}

impl dyn Error {
    /// Retrieves a reference to `T`-typed context from the error if it is available.
    pub fn context<T: Any>(&self) -> Option<&T> {
        self.provide_context(TypeId::of::<T>())?.downcast_ref::<T>()
    }
}
```